0000-richer-documents.rst

```diff
@@ -0,0 +1,300 @@
+      | TextBeside !TextDetails {-# UNPACK #-} !Int Doc  -- text s <> x
+      | Nest {-# UNPACK #-} !Int Doc                     -- nest k x
+      | Union Doc Doc                                    -- ul `union` ur
+      | NoDoc                                            -- The empty set of documents
+      | Beside Doc Bool Doc                              -- True <=> space between
+      | Above Doc Bool Doc                               -- True <=> never overlap
+
+    data SDocContext = SDC { ... }
+    newtype SDoc = SDoc { runSDoc :: SDocContext -> Doc }
+
+The exact contents of ``SDocContext`` is not very important, one just needs
+to know that it gives access to all sorts of pretty-printing and GHC
+related configuration. These two types allow us to specify how sub-documents
+should be placed relative to one another once rendered and have been great to
+describe all sorts of textual documents in GHC.
+
+On the other hand, authors of tools in the IDE/editor-integration space
+have been wondering about the possibility of enriching those error
+documents, with several applications in mind -- we will mention a few in
+the **Effects and Interactions** section. All of those applications require
+that we keep some information around, stored in good old ADTs, that we can
+freely inspect and consume: bits of expressions, types, constraints, etc.
+
+To bridge that gap, this proposal suggests that we add support for
+embedding annotations (non-textual values) directly into our documents,
+as described in the next section.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal suggests that we modify ``Doc`` to extend it with a
+constructor for embedding arbitrary values into documents, which we refer
+to as "annotations".::
+
+    data Doc a
+      = Empty
+        -- empty
+      | ...
+      | Pure a -- new! embed an annotation of type 'a'
+
+Such a document model comes with useful instances for standard
+classes, among which ``Functor``, ``Aplicative``, ``Monad``, ``Foldable``
+and ``Traversable``, giving us standard names for functions that
+transform annotations into other annotations or into whole new sub-documents.
+
+``SDoc`` would similarly be updated to take a type parameter for the
+annotation type:::
+
+    newtype SDoc a = SDoc { runSDoc :: SDocContext -> Doc a }
+
+and would come with similar instances as ``Doc`` (since with this definition,
+``SDoc ~ ReaderT SDocContext Doc``), except for ``Foldable`` and
+``Traversable``. The main addition to the public API of the ``Outputable``
+module would be a function for constructing a document from an annotation:::
+
+    embed :: a -> SDoc a
+    embed = pure
+
+We could then create an annotation type for error documents, so as to
+embed bits of expressions, types and more, in their AST form, right into
+the documents, even supporting embedding ASTs from all 3 GHC passes.::
+
+    data PassWrapper f
+      = WrapperPs (f GhcPs) -- parser AST
```

Comment:
This representation suggests that one error message might have both e.g. `HsExpr GhcPs` and `HsExpr GhcRn` in the same message. I doubt we want that. So I'm unconvinced here.

Reply:
Yes, this was just a quick idea to easily be able to embed AST fragments from all 3 phases.

On the other hand, do we really want to parametrize `ErrAnnotation` as a whole by the pass that the annotations belong to, yielding documents of type `SDoc (ErrAnnotation GhcTc)` and so on? 
We would have to make the pass "disappear" from the type at some point to be able to uniformly process documents with annotations from the different phases. One way is to wrap `ErrAnnotation`, another is to wrap the entire document. The former is almost what I suggested (except that ErrAnnotation would be the "wrappee" this time), the latter seems more intrusive since we would have to take that "wrapped document" type in many places instead of `SDoc`s.